src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs

```diff
@@ -527,10 +532,15 @@ internal bool TryGetValue(int index, out DateTimeOffset value)
 
             Debug.Assert(segment.IndexOf(JsonConstants.BackSlash) == -1);
 
-            value = default;
-            return (segment.Length <= JsonConstants.MaximumDateTimeOffsetParseLength)
-                && JsonHelpers.TryParseAsISO(segment, out value, out int bytesConsumed)
-                && segment.Length == bytesConsumed;
+            if (segment.Length > JsonConstants.MaximumDateTimeOffsetParseLength
+                || !JsonHelpers.TryParseAsISO(segment, out value, out int bytesConsumed)
+                || segment.Length != bytesConsumed)
+            {
+                value = default;
+                return false;
```

Comment:
I prefer this style over the style across the other APIs in this file. It is also consistent with the reader APIs.

So:
1) Special case  the "false" case and set value to default, otherwise return true (don't special case the "true" case).
2) Parse directly to the out parameter (`value`) rather than a temp.

Can you make the same code-pattern change across the other TryGet methods in this file. For example:
https://github.com/dotnet/corefx/blob/c63e5ebecc5449a360b6293ecdbf151bfd11a1a6/src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs#L405-L413

cc @bartonjs, if you strongly disagree. At the very least we should be consistent and we save the unnecessary assignment for the true case (and the method becomes a couple instructions smaller - [sharplab.io](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAmAjAWAChMAGAAk1wDoAhAVwDMGBTAJ2SoBVmw4BuIkUwBmCtjIBhMgG8iZeWQCWAOzhtliADZlgAe13bOrAJ4BxZnABqWuswAUKuEuUATHjDK66Tuo7IA3G2YPACVmRBcAeWVNYwBlAAdEZQAeYGM1AD4yZGYAc1RmVQBKOQVZQgUqpQYyOwBVOAYADgAFRHY2LhN2zrtcgqK4Dy8fPzhUBJHvZycAY11lZDpCl2KyADINsurqhaWV5hcyAF4TnPzC1SoAGSK8uAALUsrdmR23+UDNW1OyCYSAlen3kmAA7P9WLYgZ8AL6CYHVb6/c5uBiIOiafgfKrgsjozS5GEKeGIj6OdRaHT6QwmcxWILYByqZxuMDTMYs5HBMhhCLRWKJZJpDLMbIDK5wF5VCqfRS1OwAQkaLV6uVY3WMavsEqGHICQX1fn2y1W63Q6BxbxNh2OZwug2ud2UD2eVvkspBXyCfzRGKxxK9eIJRPdZFJYbxcChzED4aIsKAA===)).

Reply:
I think all the ones I wrote were "special case true" already, so apparently my fingers prefer that style already. :smile:

Reply:
Yep, I'll put together an update for those cases as well in that file.

Reply:
Oooh.  I see, you were asking if I agreed that "return false" should be inside the if.  No, I disagree to that.  (I read "can you make the pattern change ... such as ..." to mean "this is an example of what I want")

Only mildly, but still disagree.

I also almost always use a `tmp` so that the `out` variable doesn't leak a value in the case of an exception.

Reply:
As in an exception in `Utf8Parser.TryParse` that leaves its out parameter in an inconsistent state that we then propagate out? I'm having a hard time finding a point at which an exception would be thrown in the current JsonDocument code after value is assigned in the downstream TryParse method.

Reply:
> As in an exception in Utf8Parser.TryParse that leaves its out parameter in an inconsistent state that we then propagate out?

Yeah.  Because Utf8Parser is itself defensive in this regard there's not any obvious place that it spills (all the methods I looked at do a single assignment to the out parameter, after calculating the final answer in a local).

There are two potential problems:

1) If the caller of this method is writing to a field (e.g. element.TryGetInt32(out _numTickets)) and the implementation causes two writes to the out (of different values), a multithreaded access on _numTickets may process the incomplete answer.
2) If the caller is writing to a field, or a local within a try/catch (whose scope extends beyond the try/catch) and an exception is thrown, then the intermediate state of the out leaks and may lead to weird results (a try/finally that uses the variable in the finally, for example, incorrectly assuming it's in a coherent state).

Both of them are solved by just having a general pattern of ensuring only one write happens to the out for any public method.  (Internals/privates can be more sloppy)

Reply:
@bartonjs I can add back in the assignment to `tmp`, which should ensure going forward that `value` is never set inconsistently within JsonDocument's usage should a call into another area allow value to be set down an exceptional path that would have otherwise returned `false`.

Reply:
> Both of them are solved by just having a general pattern of ensuring only one write happens to the out for any public method

I think that is being unnecessarily defensive, but I don't feel too strongly about it.

Thanks @watfordgnf for making it consistent across the board, which was my primary concern. It helps with maintainability.

Reply:
Looking at other, existing APIs in the framework, looks like we don't always follow the "assign to temp variable" pattern to guard against exceptions.

For example:
```C#
[Fact]
public static void OutParametersNotGaurdedAgainstExceptions()
{
    ImmutableDictionary<string, int> dict = new Dictionary<string, int>
    {
        { "foo", 5 }
    }.ToImmutableDictionary();

    IDictionary<string, int> idict = dict;
    int value = -1;
    Assert.Throws<NotSupportedException>(() => idict.Remove("foo", out value));
    Assert.Equal(5, value); // It is still 5, not -1, not default/0.
}
```
The implementation (which throws on call to `Remove`):
https://github.com/dotnet/corefx/blob/05171beb453577b516caa051ce2bea2ae48d4b83/src/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs#L41-L56

Reply:
Yeah, we don't always follow the single-assignment rule.  And it usually doesn't cause problems.  But when it does, it's hard to track down.  I consider the value of stability and ease of debuggability worth the cost of the defensive behavior.

(In the cases where we don't do it, I don't think it was an intentional "save a register", but just a "this is the most obvious way to write it")